distributed/diagnostics/nvml.py

```diff
@@ -27,82 +58,97 @@ def _in_wsl():
 
 
 def init_once():
-    global nvmlInitialized, nvmlLibraryNotFound, nvmlWslInsufficientDriver, nvmlOwnerPID
+    """Idempotent (per-process) initialization of PyNVML
 
-    # nvml monitoring disabled
-    if dask.config.get("distributed.diagnostics.nvml") is False:
-        return
+    Notes
+    -----
 
-    # nvml is already initialized on this process
-    if nvmlInitialized and nvmlOwnerPID == os.getpid():
-        return
+    Modifies global variables NVML_STATE and NVML_OWNER_PID"""
+    global NVML_STATE, NVML_OWNER_PID
 
-    # nvml failed to initialize due to missing / outdated requirements
-    if pynvml is None or nvmlLibraryNotFound or nvmlWslInsufficientDriver:
+    if NVML_STATE in {
+        NVMLState.DISABLED_PYNVML_NOT_AVAILABLE,
+        NVMLState.DISABLED_CONFIG,
+        NVMLState.DISABLED_LIBRARY_NOT_FOUND,
+        NVMLState.DISABLED_WSL_INSUFFICIENT_DRIVER,
+    }:
         return
-
-    # attempt to initialize nvml
-    try:
-        pynvml.nvmlInit()
-    except (
-        pynvml.NVMLError_LibraryNotFound,
-        pynvml.NVMLError_DriverNotLoaded,
-        pynvml.NVMLError_Unknown,
-    ):
-        nvmlLibraryNotFound = True
+    elif NVML_STATE == NVMLState.INITIALIZED and NVML_OWNER_PID == os.getpid():
         return
-
-    # set a minimum driver version for WSL so we can assume certain queries work
-    if (
-        not nvmlLibraryNotFound
-        and parse_version(pynvml.nvmlSystemGetDriverVersion().decode())
-        < parse_version(minimumWslVersion)
-        and _in_wsl()
+    elif NVML_STATE == NVMLState.UNINITIALIZED and not dask.config.get(
+        "distributed.diagnostics.nvml"
     ):
-        nvmlWslInsufficientDriver = True
+        NVML_STATE = NVMLState.DISABLED_CONFIG
         return
+    elif (
+        NVML_STATE == NVMLState.INITIALIZED
+        and NVML_OWNER_PID != os.getpid()
+        or NVML_STATE == NVMLState.UNINITIALIZED
+    ):
+        try:
+            pynvml.nvmlInit()
+        except (
+            pynvml.NVMLError_LibraryNotFound,
+            pynvml.NVMLError_DriverNotLoaded,
+            pynvml.NVMLError_Unknown,
+        ):
+            NVML_STATE = NVMLState.DISABLED_LIBRARY_NOT_FOUND
+            return
+
+        if _in_wsl() and parse_version(
+            pynvml.nvmlSystemGetDriverVersion().decode()
+        ) < parse_version(MINIMUM_WSL_VERSION):
+            NVML_STATE = NVMLState.DISABLED_WSL_INSUFFICIENT_DRIVER
+            return
+        else:
+            from distributed.worker import add_gpu_metrics
 
-    # initialization was successful
-    nvmlInitialized = True
-    nvmlOwnerPID = os.getpid()
+            # initialization was successful
+            NVML_STATE = NVMLState.INITIALIZED
+            NVML_OWNER_PID = os.getpid()
+            add_gpu_metrics()
```

Comment:
What happens here if the process is a scheduler only?

Reply:
`add_gpu_metrics` does this:

```
# in worker.py (DEFAULT_METRICS) is a global var in that module
def add_gpu_metrics():
    async def gpu_metric(worker):
        result = await offload(nvml.real_time)
        return result

    DEFAULT_METRICS["gpu"] = gpu_metric

    def gpu_startup(worker):
        return nvml.one_time()

    DEFAULT_STARTUP_INFORMATION["gpu"] = gpu_startup
```

So I think this is safe if run on a process that is the scheduler (previously this same code would have been run on the scheduler process when distributed was imported)

Reply:
Yeah, ok, seems safe in that case. Thanks for detailing that.